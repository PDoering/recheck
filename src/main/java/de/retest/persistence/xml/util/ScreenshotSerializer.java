package de.retest.persistence.xml.util;

import java.io.File;

import de.retest.ui.descriptors.Element;
import de.retest.ui.descriptors.IdentifyingAttributes;
import de.retest.ui.image.ImageUtils;
import de.retest.ui.image.Screenshot;
import de.retest.util.FileUtil;

public class ScreenshotSerializer {

	public static final String EXPECTED_POSTFIX = "_expected";
	public static final String ACTUAL_POSTFIX = "_actual";

	private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger( ScreenshotSerializer.class );

	private File baseDir;

	public ScreenshotSerializer( final File baseDir ) {
		this.baseDir = baseDir;
	}

	public class DiffScreenshots {
		public final File expected;
		public final File actual;

		public DiffScreenshots( final File expected, final File actual ) {
			this.expected = expected;
			this.actual = actual;
		}

		public File getExpected() {
			return expected;
		}

		public File getActual() {
			return actual;
		}
	}

	public DiffScreenshots saveDiffScreenshots( final IdentifyingAttributes identifyingAttributes,
			final Screenshot expectedScreenshot, final Screenshot actualScreenshot ) {
		File expected = null;
		File actual = null;
		if ( expectedScreenshot != null ) {
			expected =
					saveScreenshot( expectedScreenshot, identifyingAttributes.getPath() + EXPECTED_POSTFIX + ".png" );
		}
		if ( actualScreenshot != null ) {
			actual = saveScreenshot( actualScreenshot, identifyingAttributes.getPath() + ACTUAL_POSTFIX + ".png" );
		}
		return new DiffScreenshots( expected, actual );
	}

	public void saveScreenshots( final Element descriptor ) {
		saveScreenshot( descriptor.getScreenshot(), descriptor.getIdentifyingAttributes().getPath() + ".png" );
		for ( final Element childElement : descriptor.getContainedElements() ) {
			saveScreenshots( childElement );
		}
	}

	public File saveScreenshot( final Element component ) {
		if ( component.getScreenshot() == null ) {
			return null;
		}
		return saveScreenshot( component.getScreenshot(), component.getIdentifyingAttributes().getPath() + ".png" );
	}

	public File saveScreenshot( final Screenshot image, final String path ) {
		if ( image == null ) {
			return null;
		}
		try {
			final File result = new File( baseDir, path );
			FileUtil.ensureFolder( result );
			ImageUtils.exportScreenshot( image, result );
			return result;
		} catch ( final Exception exc ) {
			logger.error( "Exception saving screenshot: ", exc );
			return null;
		}
	}

	public File getBaseDir() {
		return baseDir;
	}

	public void setBaseDir( final File baseDir ) {
		this.baseDir = baseDir;
	}
}
