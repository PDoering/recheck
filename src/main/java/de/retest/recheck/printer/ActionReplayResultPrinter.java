package de.retest.recheck.printer;

import java.util.stream.Collectors;

import de.retest.recheck.NoRecheckFileActionReplayResult;
import de.retest.recheck.ignore.ShouldIgnore;
import de.retest.recheck.report.ActionReplayResult;
import de.retest.recheck.ui.DefaultValueFinder;
import de.retest.recheck.ui.actions.ExceptionWrapper;

public class ActionReplayResultPrinter implements Printer<ActionReplayResult> {

	private final ElementDifferencePrinter printer;
	private final ShouldIgnore ignore;

	public ActionReplayResultPrinter( final DefaultValueFinder defaultValueFinder, final ShouldIgnore ignore ) {
		printer = new ElementDifferencePrinter( defaultValueFinder, ignore );
		this.ignore = ignore;
	}

	@Override
	public String toString( final ActionReplayResult difference, final String indent ) {
		final String prefix = indent + difference.getDescription() + " resulted in:\n";
		final ExceptionWrapper error = difference.getThrowableWrapper();
		if ( error != null ) {
			return prefix + indent + "\t" + error;
		}
		final Throwable targetNotFound = difference.getTargetNotFoundException();
		if ( targetNotFound != null ) {
			return prefix + indent + "\t" + targetNotFound;
		}
		if ( difference instanceof NoRecheckFileActionReplayResult ) {
			return prefix + indent + "\t" + NoRecheckFileActionReplayResult.MSG_LONG;
		}
		final String diffs = difference.getAllElementDifferences().stream() //
				.filter( diff -> !ignore.shouldIgnoreElement( diff.getElement() ) ) //
				.filter( diff -> !diff.getAttributeDifferences( ignore ).isEmpty() )
				.map( diff -> printer.toString( diff, indent + "\t" ) ) //
				.collect( Collectors.joining( "\n" ) );
		return prefix + diffs;
	}

}
