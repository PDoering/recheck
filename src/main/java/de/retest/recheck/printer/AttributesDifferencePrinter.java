package de.retest.recheck.printer;

import java.util.stream.Collectors;

import de.retest.recheck.ui.DefaultValueFinder;
import de.retest.recheck.ui.descriptors.IdentifyingAttributes;
import de.retest.recheck.ui.diff.AttributesDifference;

public class AttributesDifferencePrinter implements Printer<AttributesDifference> {

	private final AttributeDifferencePrinter delegate;

	public AttributesDifferencePrinter( final IdentifyingAttributes attributes, final DefaultValueFinder finder ) {
		delegate = new AttributeDifferencePrinter( attributes, finder );
	}

	@Override
	public String toString( final AttributesDifference difference, final String indent ) {
		return difference.getDifferences().stream() //
				.map( d -> delegate.toString( d, indent ) ) //
				.collect( Collectors.joining( "\n" ) );
	}
}
