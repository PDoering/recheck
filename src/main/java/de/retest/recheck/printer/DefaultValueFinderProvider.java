package de.retest.recheck.printer;

import de.retest.recheck.ui.DefaultValueFinder;

@FunctionalInterface
public interface DefaultValueFinderProvider {

	DefaultValueFinder findForAction( String name );

	static DefaultValueFinderProvider none() {
		final DefaultValueFinder noDefaults = ( attributes, key, value ) -> false;
		return name -> noDefaults;
	}
}
