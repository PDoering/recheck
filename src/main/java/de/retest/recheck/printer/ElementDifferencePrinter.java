package de.retest.recheck.printer;

import java.util.stream.Collectors;

import de.retest.recheck.printer.highlighting.DefaultHighlighter;
import de.retest.recheck.printer.highlighting.Highlighter;
import de.retest.recheck.ui.DefaultValueFinder;
import de.retest.recheck.ui.descriptors.IdentifyingAttributes;
import de.retest.recheck.ui.diff.ElementDifference;
import de.retest.recheck.ui.diff.InsertedDeletedElementDifference;
import de.retest.recheck.ui.diff.LeafDifference;

public class ElementDifferencePrinter implements Printer<ElementDifference> {

	private final DefaultValueFinder finder;
	private final Highlighter highlighter;

	public ElementDifferencePrinter( final DefaultValueFinder finder ) {
		this.finder = finder;
		highlighter = new DefaultHighlighter();
	}

	public ElementDifferencePrinter( final DefaultValueFinder finder, final Highlighter highlighter ) {
		this.finder = finder;
		this.highlighter = highlighter;
	}

	@Override
	public String toString( final ElementDifference difference, final String indent ) {
		return indent + createDescription( difference ) + "\n" + createDifferences( difference, indent + "\t" );
	}

	private String createDescription( final ElementDifference difference ) {
		final IdentifyingAttributes attributes = difference.getIdentifyingAttributes();
		return attributes.getType() + " (" + difference.getElement() + ") at '" + attributes.getPath() + "':";
	}

	private String createDifferences( final ElementDifference difference, final String indent ) {
		final LeafDifference identifyingAttributesDifference = difference.getIdentifyingAttributesDifference();
		if ( identifyingAttributesDifference instanceof InsertedDeletedElementDifference ) {
			final Printer<InsertedDeletedElementDifference> printer = new InsertedDeletedElementDifferencePrinter();
			return printer.toString( (InsertedDeletedElementDifference) identifyingAttributesDifference, indent );
		}
		final IdentifyingAttributes attributes = difference.getIdentifyingAttributes();
		final AttributeDifferencePrinter delegate = new AttributeDifferencePrinter( attributes, finder, highlighter );
		return difference.getAttributeDifferences().stream() //
				.map( d -> delegate.toString( d, indent ) ) //
				.collect( Collectors.joining( "\n" ) );
	}
}
