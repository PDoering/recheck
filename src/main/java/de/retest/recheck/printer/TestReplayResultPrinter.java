package de.retest.recheck.printer;

import java.util.stream.Collectors;

import de.retest.recheck.ignore.Filter;
import de.retest.recheck.report.ActionReplayResult;
import de.retest.recheck.report.TestReplayResult;
import de.retest.recheck.ui.DefaultValueFinder;

public class TestReplayResultPrinter implements Printer<TestReplayResult> {

	private final DefaultValueFinderProvider provider;
	private final Filter ignore;

	public TestReplayResultPrinter( final DefaultValueFinderProvider provider, final Filter ignore ) {
		this.provider = provider;
		this.ignore = ignore;
	}

	@Override
	public String toString( final TestReplayResult difference, final String indent ) {
		return indent + createDescription( difference ) + "\n" + createDifferences( difference, indent );
	}

	private String createDescription( final TestReplayResult result ) {
		final String name = result.getName();
		final int differences = result.getDifferences( ignore ).size();
		final int states = result.getActionReplayResults().size();
		return String.format( "Test '%s' has %d differences in %d states:", name, differences, states );
	}

	private String createDifferences( final TestReplayResult difference, final String indent ) {
		return difference.getActionReplayResults().stream() //
				.map( a -> formatAction( a, indent ) ) //
				.collect( Collectors.joining( "\n" ) );
	}

	private String formatAction( final ActionReplayResult result, final String indent ) {
		final DefaultValueFinder finder = provider.findForAction( result.getDescription() );
		final ActionReplayResultPrinter printer = new ActionReplayResultPrinter( finder, ignore );
		return printer.toString( result, indent );
	}
}
