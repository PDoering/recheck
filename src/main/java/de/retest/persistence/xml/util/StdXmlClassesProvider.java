package de.retest.persistence.xml.util;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import de.retest.elementcollection.ElementCollection;
import de.retest.persistence.xml.ReTestXmlDataContainer;
import de.retest.suite.CapturedSuite;
import de.retest.suite.ExecutableSuite;
import de.retest.ui.actions.ActionParameter;
import de.retest.ui.actions.ActionSequence;
import de.retest.ui.actions.ParameterizedAction;
import de.retest.ui.descriptors.CodeLocAttribute;
import de.retest.ui.descriptors.ContextAttribute;
import de.retest.ui.descriptors.DefaultAttribute;
import de.retest.ui.descriptors.OutlineAttribute;
import de.retest.ui.descriptors.ParameterType;
import de.retest.ui.descriptors.PathAttribute;
import de.retest.ui.descriptors.ScreenshotAttributeDifference;
import de.retest.ui.descriptors.StringAttribute;
import de.retest.ui.descriptors.SuffixAttribute;
import de.retest.ui.descriptors.TextAttribute;
import de.retest.ui.descriptors.WeightedTextAttribute;
import de.retest.ui.diff.AttributeDifference;
import de.retest.ui.diff.AttributesDifference;
import de.retest.ui.diff.DurationDifference;
import de.retest.ui.diff.ElementDifference;
import de.retest.ui.diff.IdentifyingAttributesDifference;
import de.retest.ui.diff.InsertedDeletedElementDifference;
import de.retest.ui.diff.RootElementDifference;
import de.retest.ui.diff.StateDifference;

public class StdXmlClassesProvider {

	static {
		ParameterType.registerStdParameterTypes();
	}

	private static final Class<?>[] classes = new Class<?>[] { //
			ActionParameter.class, //
			ParameterizedAction.class, //
			ActionSequence.class, //
			CapturedSuite.class, //
			ElementCollection.class, //
			ExecutableSuite.class, //
			ReTestXmlDataContainer.class, //
			de.retest.test.Test.class, //
			// we need these to handle subtypes properly: (see RET-357)
			DefaultAttribute.class, //
			PathAttribute.class, //
			SuffixAttribute.class, //
			StringAttribute.class, //
			TextAttribute.class, //
			CodeLocAttribute.class, //
			ContextAttribute.class, //
			OutlineAttribute.class, //
			WeightedTextAttribute.class, //

			DurationDifference.class, //
			AttributesDifference.class, //
			AttributeDifference.class, //
			ScreenshotAttributeDifference.class, //
			StateDifference.class, //
			IdentifyingAttributesDifference.class, //
			InsertedDeletedElementDifference.class, //
			RootElementDifference.class, //
			ElementDifference.class, //
	};

	public static Class<?>[] getXmlDataClasses( final Class<?>... additionalClazzes ) {
		return getXmlDataClasses( new HashSet<>( Arrays.asList( additionalClazzes ) ) );
	}

	public static Class<?>[] getXmlDataClasses( final Set<Class<?>> additionalClazzes ) {
		final Set<Class<?>> clazzes = new HashSet<>( Arrays.asList( classes ) );
		clazzes.addAll( additionalClazzes );
		return clazzes.toArray( new Class<?>[clazzes.size()] );
	}
}
