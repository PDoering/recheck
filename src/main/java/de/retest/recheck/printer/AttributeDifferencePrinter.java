package de.retest.recheck.printer;

import java.io.Serializable;
import java.util.List;
import java.util.stream.Collectors;

import de.retest.recheck.ui.DefaultValueFinder;
import de.retest.recheck.ui.descriptors.IdentifyingAttributes;
import de.retest.recheck.ui.diff.AttributeDifference;
import de.retest.recheck.ui.diff.ElementIdentificationWarning;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class AttributeDifferencePrinter implements Printer<AttributeDifference> {

	private static final String IS_DEFAULT = "(default or absent)";

	private static final String KEY_EXPECTED_ACTUAL_FORMAT = "%s: expected=\"%s\", actual=\"%s\"";

	private static final String WARNING_FILENAME_LINE_FORMAT = ", breaks=\"%s\"";

	private final IdentifyingAttributes attributes;
	private final DefaultValueFinder defaultProvider;

	@Override
	public String toString( final AttributeDifference difference, final String indent ) {
		if ( difference.hasElementIdentificationWarning() ) {
			return indent + format( difference ) + printWarning( difference );
		}
		return indent + format( difference );
	}

	private String format( final AttributeDifference difference ) {
		if ( isExpectedDefault( difference ) ) { // The attribute changed from default to a non-default
			return printExpectedDefaultDifference( difference );
		}
		if ( isActualDefault( difference ) ) { // The attribute changed back to default
			return printActualDefaultDifference( difference );
		}
		return printBothDifferences( difference );
	}

	private boolean isActualDefault( final AttributeDifference difference ) {
		return defaultProvider.isDefaultValue( attributes, difference.getKey(), difference.getActual() );
	}

	private String printActualDefaultDifference( final AttributeDifference difference ) {
		final String key = difference.getKey();
		final Serializable expected = difference.getExpected();
		return String.format( KEY_EXPECTED_ACTUAL_FORMAT, key, expected, IS_DEFAULT );
	}

	private boolean isExpectedDefault( final AttributeDifference difference ) {
		return difference.getExpected() == null; // We do not save defaults, thus this is null
	}

	private String printExpectedDefaultDifference( final AttributeDifference difference ) {
		final String key = difference.getKey();
		final Serializable actual = difference.getActual();
		return String.format( KEY_EXPECTED_ACTUAL_FORMAT, key, IS_DEFAULT, actual );
	}

	private String printBothDifferences( final AttributeDifference difference ) {
		final String key = difference.getKey();
		final Serializable expected = difference.getExpected();
		final Serializable actual = difference.getActual();
		return String.format( KEY_EXPECTED_ACTUAL_FORMAT, key, expected, actual );
	}

	private String printWarning( final AttributeDifference difference ) {
		final List<ElementIdentificationWarning> warnings = difference.getElementIdentificationWarnings();
		return String.format( WARNING_FILENAME_LINE_FORMAT, warnings.stream() //
				.map( this::formatWarning ) //
				.collect( Collectors.joining( ", " ) ) );
	}

	private String formatWarning( final ElementIdentificationWarning warning ) {
		return warning.getTestFileName() + ":" + warning.getTestLineNumber();
	}
}
