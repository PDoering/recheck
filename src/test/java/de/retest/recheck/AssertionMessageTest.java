package de.retest.recheck;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.io.File;
import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.Set;

import org.junit.jupiter.api.Test;

import de.retest.recheck.printer.TestReplayResultPrinter;
import de.retest.recheck.report.ActionReplayResult;
import de.retest.recheck.report.TestReplayResult;
import de.retest.recheck.ui.descriptors.Element;
import de.retest.recheck.ui.descriptors.IdentifyingAttributes;
import de.retest.recheck.ui.diff.InsertedDeletedElementDifference;
import de.retest.recheck.ui.diff.LeafDifference;

class AssertionMessageTest {

	@Test
	void no_golden_master_message_should_be_formatted_properly() throws Exception {
		final String suiteName = "SomeSuite";

		final String gmPath0 = "/gm/path/0";
		final ActionReplayResult actionReplayResult0 = mock( NoGoldenMasterActionReplayResult.class );
		when( actionReplayResult0.getGoldenMasterPath() ).thenReturn( gmPath0 );

		final String gmPath1 = "/gm/path/0";
		final ActionReplayResult actionReplayResult1 = mock( NoGoldenMasterActionReplayResult.class );
		when( actionReplayResult1.getGoldenMasterPath() ).thenReturn( gmPath1 );

		final TestReplayResult testReplayResult = new TestReplayResult( "test-name", 1 );
		testReplayResult.addAction( actionReplayResult0 );
		testReplayResult.addAction( actionReplayResult1 );

		final AssertionMessage cut = new AssertionMessage( suiteName, null, testReplayResult, null, null );

		assertThat( cut ).hasToString( "'" + suiteName + "':\n" //
				+ NoGoldenMasterActionReplayResult.MSG_LONG + "\n" //
				+ gmPath0 + "\n" //
				+ gmPath1 );
	}

	@Test
	void differences_message_should_be_formatted_properly() throws Exception {
		final String suiteName = "SomeSuite";

		final String elementPath = "foo/bar/baz";
		final IdentifyingAttributes identifyingAttributes = mock( IdentifyingAttributes.class );
		when( identifyingAttributes.getPath() ).thenReturn( elementPath );

		final Element absent = null;
		final Element present = mock( Element.class );
		when( present.getIdentifyingAttributes() ).thenReturn( identifyingAttributes );

		final InsertedDeletedElementDifference insertion =
				InsertedDeletedElementDifference.differenceFor( absent, present );
		final InsertedDeletedElementDifference deletion =
				InsertedDeletedElementDifference.differenceFor( present, absent );
		// Only inserted/deleted differences from unique differences should be accounted.
		final LeafDifference ignore = mock( LeafDifference.class );

		// Use LinkedHashSet to guarantee order.
		final Set<LeafDifference> uniqueDifferences =
				new LinkedHashSet<>( Arrays.asList( insertion, deletion, ignore ) );

		final TestReplayResult testReplayResult = new TestReplayResult( "test-name", 1 );
		testReplayResult.addAction( mock( ActionReplayResult.class ) );
		testReplayResult.addAction( mock( ActionReplayResult.class ) );

		final String allDiffs = "\tSome diff\n\tAnother diff";
		final TestReplayResultPrinter testReplayResultPrinter = mock( TestReplayResultPrinter.class );
		when( testReplayResultPrinter.toString( testReplayResult ) ).thenReturn( allDiffs );

		final String reportPath = "/report/path";
		final File resultFile = mock( File.class );
		when( resultFile.getAbsolutePath() ).thenReturn( reportPath );

		final AssertionMessage cut = new AssertionMessage( suiteName, uniqueDifferences, testReplayResult,
				testReplayResultPrinter, resultFile );

		assertThat( cut ).hasToString( "A detailed report will be created at '" + reportPath + "'. " //
				+ "You can review the details by using our CLI (https://github.com/retest/recheck.cli/) or GUI (https://retest.de/review/).\n" //
				+ "\n" //
				+ "2 check(s) in '" + suiteName + "' found the following difference(s):\n" //
				+ allDiffs //
				+ "\t" + elementPath + " was inserted!\n" //
				+ "\t" + elementPath + " was deleted!" );
	}

}
