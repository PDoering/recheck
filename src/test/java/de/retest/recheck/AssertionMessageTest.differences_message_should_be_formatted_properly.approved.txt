A detailed report will be created at '/report/path'. You can review the details by using our CLI (https://github.com/retest/recheck.cli/) or GUI (https://retest.de/review/).

2 check(s) in 'SomeSuite' found the following difference(s):
	Some diff
	Another diff
	foo/bar/baz was inserted!
	foo/bar/baz was deleted!
